### PR TITLE
WarmupTask, a mechanism to improve request latency

### PR DESCRIPTION
**Pull request by swankjesse**
We've been having with newly-started services having unacceptable
latency. This is a new mechanism to plug-in warmup tasks.

### CODE REVIEW

**Comment by swankjesse** [comment]
I was reading this Tweet repeatedly while writing this!

https://twitter.com/APStylebook/status/588063260726972416
> Warmup is one word when it's a noun, warm up is two words as a verb. #APStyleChat 
@TheSocialBlitz

File: misk-warmup/src/test/kotlin/misk/warmup/WarmupTest.kt [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+      "LoggingService startUp",
+      "LoggingWarmupTask created on warmup-0",
+      "LoggingWarmupTask warming on warmup-0",
+      "HealthChecks all passed",
+      "LoggingService shutDown",
+    )
+  }
+
+  @Test
+  fun `service is healthy even after warmup task throws`() {
```
Comment (zcross) [inline misk-warmup/src/test/kotlin/misk/warmup/WarmupTest.kt:45]: nice behavior, and nice test

File: misk-warmup/src/test/kotlin/misk/warmup/WarmupTest.kt [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    startUpAndShutDown(
+      ServiceModule<LoggingService>(),
+      WarmupModule<ThrowingWarmupTask>(),
+    )
+
+    assertThat(events).containsExactly(
```
Comment (zcross) [inline misk-warmup/src/test/kotlin/misk/warmup/WarmupTest.kt:51]: nit: assert `ERROR` level?

**Review by zcross — APPROVED** [review]
Love it, thanks for making it reusable up front; as soon as we share example impacts of using this in some of our services, I suspect others may follow suit

**Review by adrw — APPROVED** [review]
LG

File: misk-warmup/src/test/kotlin/misk/warmup/WarmupTest.kt (hunk 2 above)
Comment (swankjesse) [reply misk-warmup/src/test/kotlin/misk/warmup/WarmupTest.kt:60]: Done!

**Pull request merged**